Member binding operators (#3720)
carbon-language/carbon-lang: proposals/p3720.md

Notably, there is the process of
[instance binding](/docs/design/expressions/member_access.md#instance-binding),
that can convert a method into a bound method. This is described as an
uncustomizable process, with the members of classes being non-first-class names.

With
[proposal #3646: Tuples and tuple indexing](https://github.com/carbon-language/carbon-lang/pull/3646),
tuple indexing also uses the member-access syntax, except with numeric names for
the fields.

The currently accepted proposals for functions, most notably
[Proposal #2875: Functions, function types, and function calls](https://github.com/carbon-language/carbon-lang/pull/2875),
don't support all of the different function signatures for the `Call` interface.
For example, it does not support `addr self` or explicit compile-time
parameters. That is out of scope of this proposal, and will be addressed
separately, and means that `addr self` methods won't be considered here. The
difference between functions and methods, however, is in scope.

## Proposal

We propose that Carbon defines the compound member access operator, specifically

> `x.(y)`, to be defined in terms of rewrites to invoking an interface method,

Reviewer: "define" is used twice, maybe:

```suggestion
We propose that Carbon defines the compound member access operator, specifically
`x.(y)`, in terms of rewrites to invoking an interface method,
```

// `x.(y)` is `y.((U as ValueBind(T)).Op)(x)`
interface ValueBind(T:! type) {
  let Result:! type;
  fn Op[self: Self](x: T) -> Result;
}

// For a reference expression `x` using a binding `var x: T` and an expression `y` of type `U`,
// `x.(y)` is `*y.((U as RefBind(T)).Op)(&x)`
interface RefBind(T:! type) {
  let Result:! type;
  fn Op[self: Self](p: T*) -> Result*;
}

// For a facet value, which includes all type values, `T` and an expression `y` of type `U`,
// `T.(y)` is `y.((U as TypeBind(T)).Op)()`.
interface TypeBind(T:! type) {
  let Result:! type;
  fn Op[self: Self]() -> Result;
}
```

> **Note:** `TypeBind` is its own binding interface since the members of a type
> are defined by their values, not by their type.

The other member access operators -- `x.y`, `x->y`, and `x->(y)` -- are defined
by how they rewrite into the `x.(y)` form using these two rules:

-   `x.y` is interpreted using the existing
    [member resolution rules](/docs/design/expressions/member_access.md#member-resolution).
    For example, `x.y` is treated as `x.(T.y)` for non-type values `x` with type
    `T`. These rules are expanded to support `extend api T` declarations

>     described next.

Reviewer: ```suggestion
    `T`.
```

```carbon
let v: C = {.x = 3};
Assert((v as __ValueBind_C_F).(Call(()).Op)() == 8);
Assert((v as __ValueBind_C_Static).(Call(()).Op)() == 2);
var r: C = {.x = 4};
Assert((r as __RefBind_C_F).(Call(()).Op)() == 9);
Assert((r as __RefBind_C_Static).(Call(()).Op)() == 2);
```

How does this arise?

1. First the simple member access is resolved using the type of the receiver: \
   `v.F` -> `v.(C.F)`, `v.Static` -> `v.(C.Static)`, `r.F` -> `r.(C.F)`, `r.Static`
   -> `r.(C.Static)`. \
   Note that `C.F` is `__C_F` with type `__TypeOf_C_F`, and `C.Static` is
   `__C_Static` with type `__TypeOf_C_Static`.
2. If the left of the `.` is a value expression, the value binding operator is
   applied. Otherwise if it is a reference expression, the reference binding
   operator is applied.
3. The result of the binding has a type that implements the call interface.

> **Note:** The current wording in
> [member_access.md](/docs/design/expressions/member_access.md) says that
> `v.(C.Static)` and `r.(C.Static)` are both invalid, because they don't perform
> member name lookup, instance binding, nor impl lookup -- the `v.` and `r.`
> portions are redundant. That rule is removed by this proposal.
>

> > Instead, tools such as linters can option such code as suspicious on a

Reviewer: Feel like a word might be missing here, but also maybe "option" isn't needed?

```suggestion
> Instead, tools such as linters can flag such code as suspicious on a
```

Reviewer 2: Reopening: the same wording still seems to be present, and it reads strangely to me too.

Author: Somehow this is the result of rewrites done by our precommit?!?!

  fn Op[self: Self](p: C*) -> __RefBind_C_F* {
    return p as __RefBind_C_F*;
  }
}
```

> **Note:** `TypeBind` is used for
> [non-instance interface members](#non-instance-interface-members).

Those implementations are how we get from `__C_F` with type `__TypeOf_C_F` to
`v as __ValueBind_C_F` or `r as __RefBind_C_F`:

```carbon
// `v` is a value and so uses `ValueBind`
v.F() == v.(C.F)()
      == v.(__C_F)()
      == __C_F.((__TypeOf_C_F as ValueBind(C)).Op(v))()
      == (v as __ValueBind_C_F)()
      == (v as __ValueBind_C_F).(Call(()).Op)()

// `r` is a reference expression and so uses `RefBind`
r.F() == r.(C.F)()
      == r.(__C_F)()
      == (*__C_F.((__TypeOf_C_F as RefBind(C)).Op(&r)))()
      == (*(&r as __RefBind_C_F*))()
      == (*(&r as __RefBind_C_F*)).(Call(()).Op)()
```

> **Note:** This rewrite results in a method call, but the rewrite is only
> applied once, not applied again to the resulting method call.

The last ingredient is the implementation of the call interfaces for these bound
types.

```carbon
// Since `C.F` takes `self: Self` it can be used with both
// value and reference expressions:
impl __ValueBind_C_F as Call(()) with .Result = i32;

> impl __RefBind_C_F as Call(()) with .Result = i32;

Reviewer: You don't actually show the `Op` code here, and that might be useful to make it clear that within `Op` we need some primitive way of doing the fully-explicit and no-longer-extensible call...

Something like:

```
  fn Op[self: Self]() -> i32 {
    return call_compiler_intrinsic(C.F, self as C, ());
  }
```

And then above, you might add a line to the rewrites as:

```
      == call_compiler_intrinsic(C.F, v, ())
```

And similarly for ref, and similarly for the static below. Does that make sense?

Author: I assumed there were separate `call_compiler_intrinsic` and `method_call_compiler_intrinsic`, and used `__C_F` instead of `C.F`.

> `v.(C.Static)` and `r.(C.Static)` are both invalid, because they don't perform
> member name lookup, instance binding, nor impl lookup -- the `v.` and `r.`
> portions are redundant. That rule is removed by this proposal.
>
> Instead, tools such as linters can option such code as suspicious on a
> best-effort basis, particularly when the issue is contained in a single
> expression. Such tools may still allow code that performs the same operation
> across multiple statements, as in:
>
> ```carbon
> let M:! auto = C.Static
> v.(M)();
> r.(M)();
> ```
>
> Note that if `M` is an overloaded name, it could be an instance member in some
> cases and a non-instance member in others, depending on the arguments passed.
> This is another reason to delegate this to linters analyzing a whole
> expression on a best-effort basis, rather than a strict rule just about
> binding.

The binding operators are defined using three dedicated interfaces --
`ValueBind`, `RefBind`, and `TypeBind` --
[as defined in the "proposal" section](#proposal). These binding operations are
implemented for the types of the class members:

```carbon
impl __TypeOf_C_F as ValueBind(C)
    where .Result = __ValueBind_C_F {

>   fn Op[self: Self](x: C) -> __ValueBind_C_F {

Reviewer 2: ```suggestion
  fn Op[unused self: Self](x: C) -> __ValueBind_C_F {
```

    return x as __ValueBind_C_F;
  }
}

impl __TypeOf_C_F as RefBind(C)
    where .Result = __RefBind_C_F {

>   fn Op[self: Self](p: C*) -> __RefBind_C_F* {

Reviewer 2: ```suggestion
  fn Op[unused self: Self](p: C*) -> __RefBind_C_F* {
```

    return p as __RefBind_C_F*;
  }
}

> ```

Reviewer 2: Given an expression `e` of type `C`, I'm concerned that the type of `e.F` has two different types depending on whether `e` is a reference expression or not. Specifically, given [the behavior of indexing expressions](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/expressions/indexing.md), we don't actually know the expression category of `e` in general when type-checking a generic, even when we know that it's of type `C`. Allowing the types to differ for value versus reference binding seems like it could then break coherence, in a case like `a[b].F()`, where `var a: T` and `T:! IndexWith(typeof(b)) where .Result = C`, and we don't know whether `T impls IndirectIndexWith(typeof(b))`.

For `IndexWith` / `IndirectIndexWith` we solve this by making one interface a subtype of the other. I don't think we can do that here -- neither is more general. But I think we could require the `.Result` type to be consistent by moving it to a separate interface that both `ValueBind` and `RefBind` are subtypes of:

```carbon
interface Bind(T:! type) {
  let Result:! type;
}
interface ValueBind(T:! type) {
  extend Bind(T);
  fn Op[self: Self](p: C) -> Result;
}
interface RefBind(T:! type) {
  extend Bind(T);
  fn Op[self: Self](p: C*) -> Result*;
}
```

Does that seem feasible? This would mean that we actually don't have separate `__ValueBind_C_F` and `__RefBind_C_F` types (that wouldn't be coherent), and instead have a single `__Bind_C_F`, presumably eventually having different `ValueCall` / `RefCall` implementations (which again would need a constraint that they have the same result if both are implemented).

(I think we *could* actually make `ValueBind` be a subtype of `RefBind`, which would rule out the use case where value binding *computes* a new value rather than returning an existing one -- that is, a read-only property. That use case is already a bit shaky because while you can make `x.ComputeSize` work when `x` is a value expression, you can't make it work when `x` is a reference expression. But I think the better thing to do there is to follow what we did for indexing expressions and say that if `x` doesn't implement `RefBind`, we fall back to trying `ValueBind`, as suggested in Future Work.)

Author: Done. Would you like `ValueBind` is a subtype of `RefBind` to be listed as an alternative considered?

Author: Note that "`Result` type in a common `Bind` interface" makes "`RefBind` falls back to `ValueBind` when `RefBind` is not implemented" significantly less worrisome. Perhaps it should be added to this proposal?

Author: I added a couple of sentences to the future work section.

> **Note:** `TypeBind` is used for
> [non-instance interface members](#non-instance-interface-members).

Those implementations are how we get from `__C_F` with type `__TypeOf_C_F` to
`v as __ValueBind_C_F` or `&r as __RefBind_C_F*`:

```carbon
// `v` is a value and so uses `ValueBind`
v.F() == v.(C.F)()
      == v.(__C_F)()

>       == __C_F.((__TypeOf_C_F as ValueBind(C)).Op(v))()

Reviewer 2: ```suggestion
      == __C_F.((__TypeOf_C_F as ValueBind(C)).Op)(v)()
```

>       == (v as __ValueBind_C_F)()

Reviewer 2: In line 3 here, we go from having one `.` to having two. The second one is fine -- this is simple member access into a facet, which we said doesn't get rewritten to an interface call. But the first one seems to trigger a recursive application of the same rules. I think what we want here is:

```suggestion
v.F() == v.(C.F)()
      == v.(__C_F)()
      == method_call_compiler_intrinsic(
              <function body (__TypeOf_C_F as ValueBind(C)).Op overload 0>,
              __C_F, (v))()
      == (v as __ValueBind_C_F)()
```

... though I think it'd be clearer to describe it as you currently do, and only later say that we actually call the `ValueBind.Op` method as an intrinsic call.

Author: Added.

// `r` is a reference expression and so uses `RefBind`
r.F() == r.(C.F)()
      == r.(__C_F)()

>       == (*__C_F.((__TypeOf_C_F as RefBind(C)).Op(&r)))()

Reviewer 2: ```suggestion
      == (*__C_F.((__TypeOf_C_F as RefBind(C)).Op)(&r))()
```

      == (*(&r as __RefBind_C_F*))()
```

At this point we have resolved the binding, and are left with an expression
followed by `()`. In the first case, that expression is a value expression of
type `__Value_Bind_C_F`. In the second case, it is a reference expression of
type `__RefBind_C_F`.

The last ingredient is the implementation of the call interfaces for these bound
types.

```carbon
// Since `C.F` takes `self: Self` it can be used with both
// value and reference expressions:
impl __ValueBind_C_F as Call(()) with .Result = i32 {
  // Implementation of `Call(())` for `__RefBind_C_F` is the same.
  fn Op[self: Self]() -> i32 {
    // Calls `(self as C).(C.F)()`, but without triggering
    // binding again.
    return method_call_compiler_intrinsic(

>         <function body C.F overload 0>, self as C, ());

Reviewer 2: Should we think of this as being a call (so there are two call frames on the stack for each method call), or more as if the body of `C.F` literally appears here (with `self` bound to a different value)? Under #3833 I think I'd expect to actually see two frames on the stack at runtime with the current description, which I think is undesirable.

```suggestion
    return inlined_method_call_compiler_intrinsic(
        <function body C.F overload 0>, self as C, ());
```
might work?

  // Other implementations of `Call(())` are the same.
  fn Op[unused self: Self]() -> i32 {
    // Calls `C.Static()`, without triggering binding again.
    return call_compiler_intrinsic(
               <function body C.Static overload 0>, ());
  }
}
impl __RefBind_C_Static as Call(()) with .Result = i32;
impl __TypeOf_C_Static as Call(()) where .Result = i32;
```

**Future work**: We could reduce the number of implementations required by
making `__RefBind_*` an
[extending adapter](/docs/design/generics/details.md#extending-adapter) of the
corresponding `__ValueBind_*` class.

Going back to `v.F()` and `r.F()`, after binding the next step is to resolve the
call. As described in
[proposal #2875](https://github.com/carbon-language/carbon-lang/pull/2875), this
call is rewritten to an invocation of the `Op` method of the `Call(())`
interface, using the implementations just defined. Note:

-   Passing `*(&r as __RefBind_C_F*)` to the `self` parameter of `Call(()).Op`
    converts the reference expression to a value.
-   The `Call` interface is special. We don't
    [rewrite](#instance-interface-members) calls to `Call(__).Op` to avoid

>     infinite recursion.

Reviewer 2: Is it the `Call` interface that's special, or is it that the rewrite of function call syntax goes directly to `method_call_compiler_intrinsic` instead of rewriting to another function call? Above I suggested that the rewrite of binding syntax needs to go directly to `method_call_compiler_intrinsic` instead of going to a binding plus a function call, and I wonder if *in general* we want rewrites into calls to interface member functions to go straight to the method call intrinsic, since we know that we'll have a real function with the exact signature we're expecting sitting in the witness table for the `impl` (because otherwise we generate a thunk per #3763), and not an arbitrary callable thing for which we'd need the `Call` interface.

I think the only places we *need* a direct rewrite to `method_call_compiler_intrinsic` are in the binding and call rewrites, but it would seem more consistent if we did such a direct rewrite in all cases, and per #3833 it'd reduce the overhead of a use of an overloaded operator from three (I think) SemIR call instructions to one.

Author: I think this is a good idea, but I'm unsure of what specific changes should be made. I like the "we know that we'll have a real function with the exact signature we're expecting sitting in the witness table for the `impl` (because otherwise we generate a thunk per #3763), and not an arbitrary callable thing for which we'd need the `Call` interface" rationale. I think the "Instance interface members" section should also be updated, but it isn't immediately clear to me how it should be modified. In particular, it seems like we do want `c.(I.F)` to be a callable object when it is not immediately invoked.

Also, should the original version here be documented as an alternative considered?

Reviewer 2: Per open discussion, let's continue to treat `Call` as special for now.

Author: I will add an alternative considered for directly rewriting all calls to interface member functions to method call intrinsics.

  assert(Carbon::CallMethod(&c, q) == 5);
}
```

### Instance interface members

Instance members of an interface, such as methods, can use this framework. For
example, given these declarations:

```carbon
interface I {
  fn F[self: Self]();
}
class C {
  impl as I;
}
let c: C = {};
```

Then `I.F` is its own value with its own type:

```carbon
class __TypeOf_I_F {}
let __I_F:! __TypeOf_I_F = {};
```

That type implements `ValueBind` for any type that implements the interface `I`:

```carbon
class __ValueBind_I_F(T:! I) {
  adapt T;
}
impl forall [T:! I] __TypeOf_I_F as ValueBind(T)
    where .Result = __ValueBind_I_F(T) {
  fn Op[self: Self](x: T) -> __ValueBind_I_F(T) {
    // Valid since `__ValueBind_I_F(T)` adapts `T`:
    return x as __ValueBind_I_F(T);
  }
}
```

The actual dispatch to the `I.F` method of `C` happens in the implementation of
the `Call` interface of this adapter type that is the result of value binding.
So, this implementation of `C as I`:

```carbon
impl C as I {
  fn F[self: Self]() {
    Fanfare(self);
  }
}
```

Results in this implementation:

```carbon
impl __ValueBind_I_F(C) as Call(()) where .Result = () {
  fn Op[self: Self]() {

>     call_compiler_intrinsic(

Reviewer 2: ```suggestion
    method_call_compiler_intrinsic(
```
(For consistency with the intrinsic name used above.)
Or perhaps:
```suggestion
    inline_method_call_compiler_intrinsic(
```

    return x as __Binding_C_F;
  }
}

// Note that the `Result` type has to match, since
// it is an associated type in the `Bind(C)` interface
// that both `BindToValue(C)` and `BindToRef(C)` extend.
impl __TypeOf_C_F as BindToRef(C)
    where .Result = __Binding_C_F {
  fn Op[unused self: Self](p: C*) -> __Binding_C_F* {
    return p as __Binding_C_F*;
  }
}
```

> **Note:** `BindToType` is used for
> [non-instance interface members](#non-instance-interface-members).

Those implementations are how we get from `__C_F` with type `__TypeOf_C_F` to
`v as __Binding_C_F` or `&r as __Binding_C_F*`, conceptually following these
steps:

```carbon
// `v` is a value and so uses `BindToValue`
v.F() == v.(C.F)()
      == v.(__C_F)()
      == __C_F.((__TypeOf_C_F as BindToValue(C)).Op)(v)()
      == (v as __Binding_C_F)()

// `r` is a reference expression and so uses `BindToRef`
r.F() == r.(C.F)()
      == r.(__C_F)()
      == (*__C_F.((__TypeOf_C_F as BindToRef(C)).Op)(&r))()
      == (*(&r as __Binding_C_F*))()
```

However, to avoid recursive application of these same rules, we need to avoid
expressing this in terms of evaluating `__C_F.(`...`)`. Instead the third step

> shall instead use an intrinsic compiler primitive, as in:

Reviewer: ```suggestion
expressing this in terms of evaluating `__C_F.(`...`)`. Instead the third step
uses an intrinsic compiler primitive, as in:
```

    -   [Reference binding produces a value that wraps a pointer](#reference-binding-produces-a-value-that-wraps-a-pointer)
    -   [Separate interface for compile-time binding instead of type binding](#separate-interface-for-compile-time-binding-instead-of-type-binding)
    -   [Allow `extend api` of an incomplete type](#allow-extend-api-of-an-incomplete-type)

<!-- tocstop -->

## Abstract

Define the binding operation used to compute the result of `x.y`, `p->y`,
`x.(C.y)`, and `p->(C.y)` as calling a method from user-implementable
interfaces. Also allow classes to `extend api` other classes, for forwarding use
cases.

## Problem

What happens when binding is performed between an object instance and a member
of its type? We'd like to define the semantics in a way that is simple,
orthogonal, supports the use cases from C++, allows users to express their
intent in code in a natural and predictable way that is consistent with other
Carbon constructs, and is consistent with Carbon's goals.

Consider a class with a method and a field:

```carbon
class C {
  fn M[self: Self]();
  var f: i32;
}
var x: C = {.f = 2};
```

> The expressions `C.M` and `C.f` correspond roughly to
> [C++ pointers to members](https://en.cppreference.com/w/cpp/language/pointer#Pointers_to_members).
> They may be used to access the members of `x` using Carbon's
> [compound member syntax](/docs/design/expressions/member_access.md), as in
> `x.(C.M)` or `x.(C.f)`. What is their type? Can they be passed to a function
> separately from the instance of `C` to bind with it?

Reviewer: Maybe also mention the use case of a binding a *specific* instance? Something like:

```suggestion
The expressions `C.M` and `C.f` correspond roughly to
[C++ pointers to members](https://en.cppreference.com/w/cpp/language/pointer#Pointers_to_members).
They may be used to access the members of `x` using Carbon's
[compound member syntax](/docs/design/expressions/member_access.md), as in
`x.(C.M)` or `x.(C.f)`. What is their type? Can they be passed to a function
separately from the instance of `C` to bind with it?

The expression `x.M` on the other hand doesn't have a trivial correspondence in C++ despite being a useful to bind a specific instance and produce a stand-alone callable object. We would like a model that allows `x.M` to be meaningful in a way that is consistent with the existing meaning of `x.f` and generalizes well across different kinds of methods and callables.
```

uncustomizable process, with the members of classes being non-first-class names.

With
[proposal #3646: Tuples and tuple indexing](https://github.com/carbon-language/carbon-lang/pull/3646),
tuple indexing also uses the member-access syntax, except with numeric names for
the fields.

The currently accepted proposals for functions, most notably
[Proposal #2875: Functions, function types, and function calls](https://github.com/carbon-language/carbon-lang/pull/2875),
don't support all of the different function signatures for the `Call` interface.
For example, it does not support `addr self` or explicit compile-time
parameters. That is out of scope of this proposal, and will be addressed
separately, and means that `addr self` methods won't be considered here. The
difference between functions and methods, however, is in scope.

## Proposal

We propose that Carbon defines the compound member access operator, specifically
`x.(y)`, to be defined in terms of rewrites to invoking an interface method,
like other operators. There are three different interfaces used, depending on
whether `x` is a value expression, a reference expression, or a facet:

```carbon

> // For value `x` with type `T` and `y` of type `U`,

Reviewer: Generalizing a bit, somewhat to match my suggestion below:

```suggestion
// For a value expression `x` with type `T` and an expression `y` of type `U`,
```

// `x.(y)` is `y.((U as ValueBind(T)).Op)(x)`
interface ValueBind(T:! type) {
  let Result:! type;
  fn Op[self: Self](x: T) -> Result;
}

// For reference expression `var x: T` and `y` of type `U`,

> // `x.(y)` is `*y.((U as RefBind(T)).Op)(&x)`

Reviewer: The first thing isn't actually the reference expression... Maybe:

```suggestion
// For a reference expression `x` using a binding `var x: T` and an expression `y` of type `U`,
// `x.(y)` is `*y.((U as RefBind(T)).Op)(&x)`
```

interface RefBind(T:! type) {
  let Result:! type;
  fn Op[self: Self](p: T*) -> Result*;
}

> // For a facet value `T`, `y` of type `U`,

Reviewer: Tiny edit to match the above, but a bigger edit to make it more obvious that this includes all type values.

```suggestion
// For a facet value, which includes all type values, `T` and an expression `y` of type `U`,
```

// `T.(y)` is `y.((U as TypeBind(T)).Op)()`.
interface TypeBind(T:! type) {
  let Result:! type;
  fn Op[self: Self]() -> Result;
}
```

The other member access operators -- `x.y`, `x->y`, and `x->(y)` -- are defined
by how they rewrite into the `x.(y)` form using these two rules:

-   `x.y` is interpreted using the existing
    [member resolution rules](/docs/design/expressions/member_access.md#member-resolution).
    For example, `x.y` is treated as `x.(T.y)` for non-type values `x` with type
    `T`. These rules are expanded to support `extend api T` declarations
    described next.
    -   Simple member access of a facet `T`, as in `T.y`, is not rewritten into
        the `T.(`\_\_\_`)` form.

> -   `x->y` is interpreted as `(*x).y`.

Reviewer: Above you're distinguishing between `x.y` and `x.(y)`, but here you're not, which is a bit confusing. Maybe:

```suggestion
-   `x->y` and `x->(y)` are interpreted as `(*x).y` and `(*x).(y)` respectively.
```

2. If the left of the `.` is a value expression, the value binding operator is
   applied. Otherwise if it is a reference expression, the reference binding
   operator is applied.
3. The result of the binding has a type that implements the call interface.

> **Note:** The current wording in
> [member_access.md](/docs/design/expressions/member_access.md) says that
> `v.(C.Static)` and `r.(C.Static)` are both invalid, because they don't perform
> member name lookup, instance binding, nor impl lookup -- the `v.` and `r.`
> portions are redundant. That rule is now a best-effort diagnostic and should
> not trigger as the result of rewrites performed by the compiler or when the
> problem is not contained in a single expression. For example, it would be
> reasonable for the compiler to allow:
>
> ```carbon
> let M:! auto = C.Static
> v.(M)();
> r.(M)();
> ```
>
> Note that with overloaded names, the `M` might be an instance member in some
> cases and a non-instance member in others, depending on the arguments passed.
> Since the call happens after binding, this gives another reason to consider
> this a best-effort error on a whole expression, rather than a strict rule just

> > about binding.

Reviewer: I don't think we have a good concept of a "best-effort diagnostic" or a "best-effort error".

I think I'd prefer to just have this proposal remove all discussion of these being invalid and make them valid. We can of course still have linters or tooling that suggest removing redundant and unnecessary code, but as you point out, each aspect of this will now have *some* use cases and it will only be when combined in a way that doesn't use any of these things that there will be a simplification. At that point, I think we should just have the design focus on the validity.

Author: I changed it so the rule is removed, and changed the discussion to be about linting.

The binding operators are defined using three dedicated interfaces:

```carbon
// For value `x` with type `T` and `y` of type `U`,
// `x.(y)` is `y.((U as ValueBind(T)).Op)(x)`
interface ValueBind(T:! type) {
  let Result:! type;
  fn Op[self: Self](x: T) -> Result;
}

// For reference expression `var x: T` and `y` of type `U`,
// `x.(y)` is `*y.((U as RefBind(T)).Op)(&x)`
interface RefBind(T:! type) {
  let Result:! type;
  fn Op[self: Self](p: T*) -> Result*;
}

// For a facet value `T`, `y` of type `U`,
// `T.(y)` is `y.((U as TypeBind(T)).Op)()`.
interface TypeBind(T:! type) {
  let Result:! type;
  fn Op[self: Self]() -> Result;
}
```

These binding operations are implemented for the types of the class members:

```carbon
impl __TypeOf_C_F as ValueBind(C)
    where .Result = __ValueBind_C_F {
  fn Op[self: Self](x: C) -> __ValueBind_C_F {
    return x as __ValueBind_C_F;
  }
}

impl __TypeOf_C_F as RefBind(C)
    where .Result = __RefBind_C_F {
  fn Op[self: Self](p: C*) -> __RefBind_C_F* {
    return p as __RefBind_C_F*;
  }
}

> ```

Reviewer: This section is duplicated from outside the details. Could you instead reference the prior section here and not duplicate?

Author: I'm not sure what you mean. What is duplicated and where else does it appear? Are you talking about the definitions of `ValueBind`, `RefBind`, and `TypeBind` interfaces? I have removed the duplication of those, but it does not appear that the implementations of those interfaces are duplicated.

> **Note:** `TypeBind` is used for
> [non-instance interface members](#non-instance-interface-members).

Those implementations are how we get from `__C_F` with type `__TypeOf_C_F` to
`v as __ValueBind_C_F` or `r as __RefBind_C_F`:

```carbon
// `v` is a value and so uses `ValueBind`
v.F() == v.(C.F)()
      == v.(__C_F)()
      == __C_F.((__TypeOf_C_F as ValueBind(C)).Op(v))()
      == (v as __ValueBind_C_F)()
      == (v as __ValueBind_C_F).(Call(()).Op)()

> // `v` is a reference expression and so uses `RefBind`

Reviewer: ```suggestion
// `r` is a reference expression and so uses `RefBind`
```

r.F() == r.(C.F)()
      == r.(__C_F)()

>       == (*__C_F.((__TypeOf_C_F as RefBind(C)).Op(&v)))()

Reviewer: ```suggestion
      == (*__C_F.((__TypeOf_C_F as RefBind(C)).Op(&r)))()
```

      == (*(&v as __RefBind_C_F*))()
      == (*(&v as __RefBind_C_F*)).(Call(()).Op)()
```

> **Note:** This rewrite results in a method call, but the rewrite is only
> applied once, not applied again to the resulting method call.

The last ingredient is the implementation of the call interfaces for these bound
types.

```carbon
// Since `C.F` takes `self: Self` it can be used with both
// value and reference expressions:
impl __ValueBind_C_F as Call(()) with .Result = i32;
impl __RefBind_C_F as Call(()) with .Result = i32;

// `C.Static` works the same as `C.F`, except it also
// implements the call interfaces on `__TypeOf_C_Static`.
// This allows `C.Static()` to work, in addition to
// `v.Static()` and `r.Static()`.
impl __ValueBind_C_Static as Call(()) with .Result = i32;
impl __RefBind_C_Static as Call(()) with .Result = i32;
impl __TypeOf_C_Static as Call(()) where .Result = i32;
```

**Note**: We could reduce the number of implementations required by making
`__RefBind_*` an
[extending adapter](/docs/design/generics/details.md#extending-adapter) of the

> corresponding `__ValueBind_*` class.

Reviewer: Maybe say a bit more about the plans here?

IE, is this just an implementation detail of the prelude? Or is it future work to decide ? Or a rejected alternative (for now)?

Author: Marked as future work.

  // becomes: `*f.((F as RefBind(T)).Op)(&*x)`
  // The constraint `F` says the return type of
  // `(F as RefBind(T)).Op` is `i32*`, which is
  // dereferenced to get an `i32` reference expression
  // which may then be assigned.
  x->(f) = y;
}

class C {
  var m: i32;
  var n: i32;
}
var c: C = {.m = 5, .n = 6};
Assert(GetField(c, C.m) == 5);
Assert(GetField(c, C.n) == 6);
SetField(&c, C.m, 42);
SetField(&c, C.n, 12);
Assert(GetField(c, C.m) == 42);
Assert(GetField(c, C.n) == 12);
```

### C++ pointer to member

In [the generic type of member section](#generic-type-of-a-class-member), the
names of members, such as `D.K`, `X.B`, `X.V`, and `C.n`, refer to zero-sized /
stateless objects where all the offset information is encoded in the type.
However, the definitions of `CallMethod`, `SetField`, and `GetField` do not
depend on that fact and will be usable with objects, such as C++
pointers-to-members, that include the offset information in the runtime object
state. So we can define binding implementations for them so that they may be
used with Carbon's `.(`**`)` and `->(`**`)` operators.

For example, this is how I would expect C++ code to call the above Carbon

> functions:

Reviewer: (just to match the voice elsewhere)
```suggestion
For example, this is how we expect C++ code to call the above Carbon
functions:
```

}
```

### Instance interface members

Instance members of an interface, such as methods, can use this framework. For
example, given these declarations:

```carbon
interface I {
  fn F[self: Self]();
}
class C {
  impl as I;
}
let c: C = {};
```

Then `I.F` is its own value with its own type:

```carbon
class __TypeOf_I_F {}
let __I_F:! __TypeOf_I_F = {};
```

That type implements `ValueBind` for any type that implements the interface `I`:

```carbon
class __ValueBind_I_F(T:! I) {
  adapt T;
}
impl forall [T:! I] __TypeOf_I_F as ValueBind(T)
    where .Result = __ValueBind_I_F(T) {
  fn Op[self: Self](x: T) -> __ValueBind_I_F(T) {
    // Valid since `__ValueBind_I_F(T)` adapts `T`:
    return x as __ValueBind_I_F(T);
  }
}
```

The actual dispatch to the `I.F` method of `C` happens in the implementation of
the `Call` interface of this adapter type that is the result of value binding.
So, this implementation of `C as I`:

```carbon
impl C as I {
  fn F[self: Self]() {
    Fanfare(self);
  }
}
```

Results in this implementation:

```carbon
impl __ValueBind_I_F(C) as Call(()) where .Result = () {
  fn Op[self: Self]() {
    let __self: C = self as C;
    Fanfare(__self);

>   }

Reviewer: It's a bit odd that we have the somewhat inlined body of the `impl` above here...

I wonder if there is a more elegant way to describe this that is equivalent -- maybe `Op[self: C]()` would work? I feel like we really want, at some point, the parameter space and body to be *identical* to what the user wrote so that there is a pretty bright line separation between the synthesized layers and the code the user can see.

WDYT?

Author: `Op[self: C]()` is not currently allowed, to my knowledge, since `Self` -> `C` is an explicit conversion.

I think there is some expected rewriting of the users code as we express it using things that are more primitive.

Another approach would be:
```
fn __C_I_F(__self: Self) {
  Fanfare(__self);
}

impl __ValueBind_I_F(C) as Call(()) where .Result = () {
  fn Op[self: Self]() {
    let __self: C = self as C;
    __C_I_F(__self);
  }
}
```
The advantage of this is the code below code call `__C_I_F` instead of `Fanfare`.

Reviewer: I started with a suggestion to tweak this a bit, but talking with @zygoloid about it I think it was a bad suggestion...

He observes that I think we can do better than this -- we already have a function definition which accepts a `self` of type `C` and has the body we want, and we have a way of naming it (now): `(C as I).F`. And above, I suggested that we need the compiler-intrinsic-style of primitive call (analogous to the compiler intrinsic for the offset-pointer formation etc). Combined, maybe we could adjust the body of `Op` to:

```suggestion
  fn Op[self: Self]() {
    call_compiler_intrinsic((C as I).F, self as C, ());
  }
```

}
```

A call such as `c.(I.F)()` goes through these rewrites:

```carbon
c.(I.F)() == c.(__I_F)()
          == __I_F.((__TypeOf_I_F as ValueBind(C)).Op)(c)()
          == (c as __ValueBind_I_F(C))()
          == ((c as __ValueBind_I_F(C)) as Call(())).Op()

>           == Fanfare(c)

Reviewer: This last step seems a bit odd -- it isn't a rewrite, and is the body of code the user wrote. I feel like I expected the rewrites to stop at the previous line...

Author: Changed it to text explanation afterwards.

```

### Non-instance interface members

Non-instance members use the `TypeBind` interface instead. For example, if `G`
is a non-instance function of an interface `J`:

```carbon
interface J {
  fn G();
}
impl C as J;
```

Again the member is given its own type and value:

```carbon
class __TypeOf_J_G {}
let __J_G:! __TypeOf_J_G = {};
```

Since this is a non-instance member, this type implements `TypeBind` instead of
`ValueBind`:

```carbon
class __TypeBind_J_G(T:! J) {}
impl forall [T:! J] __TypeOf_J_G as TypeBind(T)
    where .Result = __TypeBind_J_G(T) {
  fn Op[self: Self]() -> __TypeBind_J_G(T) {
    return {};
  }
}
```

So, this implementation of `C as J`:

```carbon
impl C as J {
  fn G() {
    Fireworks();
  }
}
```

Results in this implementation:

```carbon
impl __TypeBind_J_G(C) as Call(()) where .Result = () {
  fn Op[self: Self]() {
    Fireworks();
  }
}
```

A call such as `C.(J.G)()` goes through these rewrites:

```carbon
C.(J.G)() == C.(__J_G)()
          == __J_G.((__TypeOf_J_G as TypeBind(C)).Op)()()
          == ({} as __TypeBind_J_G(C))()
          == (({} as __TypeBind_J_G(C)) as Call(())).Op()

>           == Fireworks()

Reviewer: Similar to the above, the last line seems odd / different.

Author: Changed.

-   `extend impl as I` adds lookup into the implementation of an interface `I`.
-   `extend base: B` adds lookup into the base class `B`.
-   `extend adapt C` adds lookup into the adapted class `C`.
-   [mixins are also planning](/proposals/p2760.md#future-work-mixins) on using
    an `extend` syntax

The natural choice is to add another way of using `extend` with consistent
lookup rules as the other uses of `extend`, with conflicts handled as determined
by leads issues
[#2355](https://github.com/carbon-language/carbon-lang/issues/2355) and
[#2745](https://github.com/carbon-language/carbon-lang/issues/2745). I propose
`extend api T`:

```carbon
class Box(T:! type) {
  var ptr: T*;
  extend api T;
}
```

This means that lookup into `Box(T)` also looks in the type `T` for members. In
this way, `b.Search` will find `b.(String.Search)`. The general rule for
`extend`, which we apply here as well, is that if lookup into `Box(T)` succeeds,
then that result is used and no lookup into `T` is performed. The extended type
must be complete. If a class uses `extend` more than once, and finds the same
name more than once, that is an ambiguity error that needs to be resolved by

> qualifying the name on use.

Reviewer: I'd like us to take a close look at this approach and see whether it is the right approach in the context of API evolution. I'm not sure it is, but I'm also not sure it isn't.

But I don't think that should block this proposal one way or the other because this is already a fundamental aspect of `extend adapt`. So it'll be a re-visit regardless, and I'm happy for it to not block anything. The goal should be to take a look across our `extend` usages and see what (if any) adjustments need to be made for evolution.

Author: Added a note to look at this in the successor proposal.

> **Note:** This is an alternative to defining an `alias` for each member of the
> extended type. This avoids having to repeat them, which is both lengthy and
> could get out of sync as the class evolves. The `extend api` approach works
> even if, as in this example, we don't know the names of the members of the
> type being extended.

Like other uses of `extend`, an `extend api` declaration do not have access
control modifiers and only operate on public names.

> **Future:** We might want `extend api` to also get (some) interface

> > implementations, like `extend base` does.

Reviewer: I feel like this is pretty important -- both `extend base` and `extend adapt` provide this, and so it seems especially likely to be needed here too.

Does it make sense to either add this now?

Or if you end up factoring `extend api` out into its own proposal, maybe it can go there? (Not advocating either way based on this point FWIW.)

Author: Added a note to look at this in the successor proposal.

    // to do reference binding not value binding.
  }
}

impl forall [T:! type, U:! RefBind(T)]
    U as RefBind(Box(T)) where .Result = U.Result {
  fn Op[self: Self](p: Box(T)*) -> Result* {
    return self.Op(p->ptr);
    // Or equivalently:
    // return p->ptr->(self);
  }
}
```

A few observations:

-   These declarations use `Box` to satisfy the orphan rule, and so this
    approach only works with `Box(T)` values, not types that can implicitly
    convert to `Box(T)`.
-   The implementation of the `Op` method is where we follow the pointer member
    `ptr` of `Box(T)` to get a `T` value that is compatible with the member
    being bound. This resolves the type mismatch that is introduced by allowing
    name resolution to find another type's members.
-   We have to be a little careful in the implementation of `ValueBind(Box(T))`
    to still use value binding even when we get a reference expression from
    dereferencing the pointer `ptr`.

With these two ingredients, `b.Search(32)` is equivalent to
`b.(String.Search)(32)`, which is then equivalent to
`b.ptr->(String.Search)(32)` or `b.ptr->Search(32)` (since `b.ptr` has type
`String*`).

> ### Other uses of `extend api`

Reviewer: Somewhat want to check my understanding first ...

It seems like this primitive could also be used to replicate the effects of all the other uses of `extend`. Basically, other uses of `extend` can be modeled with this plus potentially some additional members, implicit (or explicit) conversions, and implementations of the bind APIs. I think even inheritance can be fully represented with this model, including dynamic dispatch of virtual functions by providing special forms of the method objects?

If I've understood this correctly, I think that's really interesting and something you should document. And if we go this route, I'd actually suggest to pivot our design of other forms of `extend` as fundamentally syntactic sugar around common, expected uses of `extend api`. It feels like this gives us a much simpler, smaller atom of core language logic that everything else can build on.

Really interested if I've gotten this right or not.

However, I think *this* may be a big enough textual change to maybe factor this into a separate proposal to unblock landing the rest of bindings.

I would leave a future work section that basically hints at "we should explore this, because now that we can customize bindings, we could potentially do much more interesting things around API extension". Does that sound reasonable?

Author: Added a note to look at this in the successor proposal.

Rather than manually individually aliasing the members of `A` in `C`, to make
them available as part of the API of `C`, so that `C(i32).F()` is valid, an
`extend api` declaration includes all of them at once.

Another use case is that an interface can `extend api` of another interface. In
this example,

```carbon
interface A {
  fn F();
}
interface B {
  extend api A;
  fn G();
}
```

`B.F` would be an alias for `A.F`, but without any implied
`require Self impls A`, in contrast with a plain `extend A`. So `extend I;` is
equivalent to `require Self impls I; extend api I;`.

Lastly, an interface could `extend api` a class. This could be done to add
something that acts like `final` functions to the interface, using
[extension methods](https://github.com/carbon-language/carbon-lang/issues/1345),
as in:

```carbon
class Helpers {
  fn F[T:! type, self: T]() { DoStuffWith(self); }
  // ...
}
interface Iface {
  extend api Helpers;
  fn G();
  // ...
}
class C {
  extend impl as Iface;
}
fn Test(c: C) {
  // Calls Helpers.F from extended class Helpers
  c.F();
}
```

Unlike `final` functions in an interface, this approach defines all the helpers

> in a separate entity that could be used to extend more than one type.

Reviewer: While I see how this could work, it seems very circuitous...

Couldn't the same effect (even the factoring) be achieved by extending an interface with a set of final methods?

Or have I misunderstood something here?

Author: If `Helpers` had 20 functions, each interface that used it would have to write 20 final functions instead of one `extend api`.

Author: Looking at this again, you are correct. This can indeed be achieved by extending an interface with a set of final methods. I have made this change in #3802.

### C++ operator overloading

C++ does not support a feature like `extend api` nor does it support customizing
the behavior of `x.y`. It does support customizing the behavior of `operator*`
and `operator->` which is frequently used to support smart pointers and
iterators. There is, however, nothing restricting the implementations of those
two operators to be consistent, so that `(*x).y` and `x->y` are the same.

Carbon instead will only have a single interface for customizing dereference,
corresponding to `operator*` not `operator->`. All uses of `x->y` will be
rewritten to use `(*x).y` instead. This may cause some friction when porting C++
code where those operators are not consistent. If the C++ code is just missing
the definition of `operator*` corresponding to an `operator->`, a workaround
would be just to define `operator*`.

Other cases of divergence between those operators should be rare, since that is

> both surprising to users and for the common case of iterators, violate the C++

Reviewer: ```suggestion
both surprising to users and for the common case of iterators, violates the C++
```

  read_property area -> f64 {
    return Pi * self.radius * self.radius;
  }
}

let c: Circle = {.radius = 2};
Assert(NearlyEqual(c.area, 4 * Pi));
```

In this example, the binding of `c` of type `Circle` to `Circle.area` would
perform the computation and return the result as an `f64`.

If there was some way to customize the result of binding, this could be extended
to support other kinds of properties, such as mutable properties that use `get`
and `set` methods to access and mutate the value. The main obstacle to any
support for properties with binding is how the customization would be done. The
most natural way to support this customization would be to have multiple
interfaces. The compiler would try them in a specified order and use the one it
found first. This has the downside of the possibility of different behavior in a
checked generic context where only some of the implementations are visible.

## Rationale

This proposal is about:

-   Orthogonality: separating the binding process as a separate and independent
    step of using the members of a type.
-   Being consistent with our overall strategy for defining operators in terms
    of interface implementations.

> -   Allows binding-related functionality to be define through

Reviewer: ```suggestion
-   Allows binding-related functionality to be defined through
```

    calling them. This comes up, for example, with C++ types that define
    `operator()`.
-   We want the proposed behavior when binding a [field](#data-fields).

As a result, it would be better to evaluate this alternative later as part of
considering mutation in calls.

### Separate interface for compile-time binding instead of type binding

The first proposed way to handle
[non-instance interface members](#non-instance-interface-members) was in
[the #typesystem channel on Discord on 2024-03-07](https://discord.com/channels/655572317891461132/708431657849585705/1215405895752618134).
The suggestion was to have a `CompileBind` interface used for any compile-time
value to the left of the `.`. It would have access to the value, which is needed
when accessing the members of a type.

We eventually concluded that the special treatment was specifically needed for
types, not all compile-time values. The
[insight](https://docs.google.com/document/d/1s3mMCupmuSpWOFJGnvjoElcBIe2aoaysTIdyczvKX84/edit?resourcekey=0-G095Wc3sR6pW1hLJbGgE0g&tab=t.0#heading=h.1k8r5fhfwyh6)
was that types are special because their members are defined by their values,
not by their type

> ([which is always `type`](https://github.com/carbon-language/carbon-lang/pull/2360)).

Reviewer: I think it would be worth lifting this up to where you introduce `TypeBind` and expanding it a bit so that you front-load the core idea here.

Author: Added a note to the "Proposal" section.

interfaces. Also allow classes to `extend api` other classes, for forwarding use
cases.

## Problem

What happens when binding is performed between an object instance and a member
of its type? We'd like to define the semantics in a way that is simple,
orthogonal, supports the use cases from C++, allows users to express their
intent in code in a natural and predictable way that is consistent with other
Carbon constructs, and is consistent with Carbon's goals.

Consider a class with a method and a field:

```carbon
class C {
  fn M[self: Self]();
  var f: i32;
}
var x: C = {.f = 2};
```

The expressions `C.M` and `C.f` correspond roughly to
[C++ pointers to members](https://en.cppreference.com/w/cpp/language/pointer#Pointers_to_members).
They may be used to access the members of `x` using Carbon's
[compound member syntax](/docs/design/expressions/member_access.md), as in
`x.(C.M)` or `x.(C.f)`. What is their type? Can they be passed to a function
separately from the instance of `C` to bind with it?

The other issue is how we clearly delineate the `self` associated with a method
signature as separate from the `self` of function values.

> ## Background

Reviewer 3: C++ currently supports both `operator->` and `operator*` overloading. It sounds as if `RefBind` and `ValueBind` are both variants on `operator*`. Should this proposal comment on how migration/interop with those might work?

If I understand correctly here, `x->y` is rewritten to `(*x).y` and then `RefBind` would be called, but `operator*` and `operator->` cannot be separately overloaded as in C++. If that's the case, perhaps it should be covered as an alternative, maybe covering the implications to how a migration might occur?

Note, as just one small example, `clang::DeclListNode` provides divergent implementations:

```
    reference operator*() const {
      assert(Ptr && "dereferencing end() iterator");
      if (DeclListNode *CurNode = Ptr.dyn_cast<DeclListNode*>())
        return CurNode->D;
      return Ptr.get<NamedDecl*>();
    }
    void operator->() const { } // Unsupported.
```
https://clang.llvm.org/doxygen/DeclBase_8h_source.html

My impression here is that if wrapped by Carbon, `node->member` would then desugar to `(*node).member` and be valid, rather than invalid.

Reviewer 2: The various kinds of decl iterator in that file are all really weird and kind of broken. `*it` returns a `Decl*`, so `it->blah` really ought to not work, but instead, for most of them `operator->` is defined such that `it->blah` means the same thing as `(*it)->blah`. The example you picked out is the least weird of the bunch! :)

I think I'm fine that wrapping this in Carbon would disallow using `it->blah`. Especially if we have a convenient double-dereference syntax to use instead. (We talked about `it-->blah` at one point.)

Author: I've added a section on this topic. That specific example, if I am not mistaken, was discussed in open discussion: https://docs.google.com/document/d/1s3mMCupmuSpWOFJGnvjoElcBIe2aoaysTIdyczvKX84/edit?resourcekey=0-G095Wc3sR6pW1hLJbGgE0g&tab=t.0#heading=h.5vj8ohrvqjqh

that can convert a method into a bound method. This is described as an
uncustomizable process, with the members of classes being non-first-class names.

With
[proposal #3646: Tuples and tuple indexing](https://github.com/carbon-language/carbon-lang/pull/3646),
tuple indexing also uses the member-access syntax, except with numeric names for
the fields.

The currently accepted proposals for functions, most notably
[Proposal #2875: Functions, function types, and function calls](https://github.com/carbon-language/carbon-lang/pull/2875),
don't support all of the different function signatures for the `Call` interface.
For example, it does not support `addr self` or explicit compile-time
parameters. That is out of scope of this proposal, and will be addressed
separately, and means that `addr self` methods won't be considered here. The
difference between functions and methods, however, is in scope.

## Proposal

We propose that Carbon defines the compound member access operator, specifically
`x.(y)`, to be defined as in terms of rewrites to invoking an interface method,
like other operators. There are two different interfaces used, depending on

> whether `x` is a value expression or a reference expression:

Reviewer 3: Maybe I'm missing something about how these interfaces work, so I'd like to ask a question that's been bouncing in my head; suppose I have:

```
// Class to target with calls.
class D {}

// Declare a class which requires C -> D conversions.
class C {
  fn Foo[self: D]();
  fn Bar[addr self: D*]();

  // Apologies if I get my syntax wrong here: trying to support RefBind of C -> D.
  extend impl as RefBind(D) where .Result = D {
    fn Op[self: Self](p: C*) -> D* {
      return p as D*;
    }
  }
}

fn Run() {
  // I assume this succeeds because ValueBind isn't checked at this point;
  // although, maybe I should ask, maybe it's worth diagnosing that
  // the Foo function requires a missing ValueBind?
  var c: C;
  // I assume this uses RefBind.
  c.Bar();
  // This fails due to lack of ValueBind, right? Could it succeed using RefBind?
  c.Foo();
}
```

Even if ValueBind and RefBind can be provided separately, should there be an interface that combines both, similar to OrderedWith?

Author: (I wrote this up, but didn't reply before zygoloid@. I think the two answers agree on their overlap, but perhaps one answer will be clearer for you.)

`RefBind` (and `ValueBind`) is not about binding between two classes, but between classes and members.

`var c: C;` doesn't use `.` or `->`, so never does any binding.

`c.Bar();` does the following steps:
* Does name resolution on `Bar` to rewrite this to `c.(C.Bar)();`
* Does binding. `C.Bar` implements `RefBind(D)` and `RefBind(T)` for any `T` such that `T*` implicitly converts to `D*`. So if binding is going to succeed, we need another implementation that says "things that implement `RefBind(D)` also implement `RefBind(C)`." That would look like:
```
impl forall [T:! RefBind(D)]
    T as RefBind(C) where .Result = T.Result {
  fn Op[self: Self](p: C*) -> Result* {
    // p as D* isn't actually legal, but assuming it was for some reason
    return self.Op(p as D*);
    // Or equivalently:
    // return (p as D*)->(self);
  }
}
```
* The result of this binding is equivalent to `(&c as D*)->(C.Bar)`. It does implement the call operator.

The `c.Foo()` example proceeds in the same way. It still uses `RefBind`, since `c` is a reference expression. And in fact the result will still implement the call operator. That last step is where the reference expression is converted to a value in order to call `C.Foo`.

If you wrote:
```
let l: C;
l.Bar();
l.Foo();
```
then it would try to use `ValueBind`. Without a custom implementation, this would fail, since `C.Bar` doesn't implement `ValueBind` at all, and `C.Foo` only implements `ValueBind(D)`.

Unfortunately, I don't see a clear way of using a single function to implement both `ValueBind.Op` and `RefBind.Op`; so I don't think we are in a good position to provide an interface that can be implemented to get both (without having two functions to implement).

An alternative we could consider is having a single `Bind` interface replace both `ValueBind` and `RefBind`; with separate value and reference binding methods. However, we already have types, like `typeof(C.Bar)` that only want to implement one and not the other, so it doesn't seem viable.

Reviewer 2: I think both `c.Bar()` and `c.Foo()` use `RefBind`, because `c` is a reference expression.

I think the `impl` you want inside `C` is:

```carbon
impl forall [T:! RefBind(D)] T as RefBind(C) where . Result = T.Result {
  fn Op[self: T](p: C*) -> T.Result* {
    return &(p as D*)->(self);
  }
}
```

... and I think that supports both calls here. But it would not permit

```carbon
let c: C = ...;
// Error, no ValueBind impl.
c.Foo();
```

Reviewer 3: I wonder if we should do more for interfaces that compose separate interfaces, but I expect that's out of scope here.

```carbon
// For value `x` with type `T` and `y` of type `U`,
// `x.(y)` is `y.((U as ValueBind(T)).Op)(x)`
interface ValueBind(T:! type) {
  let Result:! type;
  fn Op[self: Self](x: T) -> Result;
}

// For reference expression `var x: T` and `y` of type `U`,
// `x.(y)` is `*y.((U as RefBind(T)).Op)(&x)`
interface RefBind(T:! type) {
  let Result:! type;
  fn Op[self: Self](p: T*) -> Result*;
}
```

The other member access operators -- `x.y`, `x->y`, and `x->(y)` -- are defined
by how they rewrite into the `x.(y)` form using these two rules:

-   `x.y` is interpreted using the existing
    [member resolution rules](/docs/design/expressions/member_access.md#member-resolution).
    For example, `x.y` is treated as `x.(T.y)` for non-type values `x` with type
    `T`. These rules are expanded to support `extend api T` declarations
    described next.
-   `x->y` is interpreted as `(*x).y`.

In addition, we propose adding a declaration in a class body to find names
defined in another class, `extend api T`:

```carbon
class Extended {
  fn F[self: Self]();
  fn G[self: Self]();
}

class Extending {
  fn G[self: Self]();
  extend api Extended;
}

> var e: Extended

Reviewer 3: ```suggestion
var e: Extending;
```

I'm trying to understand this example. Is the type incorrect?

```

This means that lookup into `Extending` also looks in the type `Extended` for

> members. In this way, `e.F` will find `e.(Extended.F)`. `extend api T` is

Reviewer 3: I'm a little confused here. Wouldn't `e.F()` be invalid because `Self` would be `Extended` while `e` is `Extending` [assuming the type is a typo]? No conversion is established; maybe you can elaborate on why this example works? Or if it doesn't, maybe it would help to explicitly state that calling `e.F()` would be invalid without the "member forwarding" wrappers?

Author: You are correct that `e.F()` is invalid unless there is a custom binding or implicit conversion. I've updated the text to explain that and include an implicit conversion.

> equivalent to defining an `alias` for every member of `T` that doesn't have a

Reviewer 3: Although the example focuses on public instance methods, I'm wondering how this interacts with other features. It sounds like a pretty complex feature and it'd help me to see more detail. For example:

- How does this interact with access control? Does `extend api` only support public APIs, not private or protected? Since this is declaring names, can I do a `protected extend api` and have everything aliased as protected instead of public?

- How does this work with non-instance methods? Does `extend api` still `alias` things like factory functions? (note, I could see this either way -- avoid it because constructors would be creating the wrong type, or embrace it for other non-instance functions, acknowledging that classes need to be pretty careful about using the feature and hiding constructors on extended classes).

- Is `extend api` valid on any type, including structs and tuples? If I `extend api (i32,)` or `extend api {.a: i32}`, what do each expose?

Reviewer 3: As I was thinking about special type attributes, I assume `extend base` is aliased without inheriting. But the destructor is probably aliased if not explicitly provided; would that result in a member's destructor being called twice?

Author: * Independent of this proposal, we currently do not allow access control keywords before `extend`, and this proposal isn't intending to change that. My expectation is that `extend` only adds aliases for public APIs. It might make sense for it to do a bit more if there is some sort of `friend` relationship between the types that would allow more access.
* My assumption was this was about visible names, and so would include fields, non-instance methods and so on. However, the only use cases I have is for instance methods and fields, so I'd be fine restricting to those.
* `extend api` on structs and tuples should not be a special case. `extend api {.a: i32}` would add an alias `alias a = {.a: i32}.a`. This would make sense if the type has an implicit conversion to `{.a: i32}`.
* I don't think destructors have names, and so can't be aliased, and I don't see any situation in which some destructor could be called twice by the introduction of this feature. Perhaps you could clarify the concern with a concrete example where you think this would be a problem?

Reviewer 3: To be sure, where you say you expect behaviors, do you plan to update the proposal to have it be specific? I'm not seeing relevant changes, but I'd expect the proposal for `extend api` to cover these.

> * Independent of this proposal, we currently do not allow access control keywords before `extend`, and this proposal isn't intending to change that.

Okay. Note, I think this is somewhat separate from other points, but I do think `protected` would probably be useful for composition, since it's extending an API in a way that's exposed to child classes.

> * `extend api` on structs and tuples should not be a special case. `extend api {.a: i32}` would add an alias `alias a = {.a: i32}.a`. This would make sense if the type has an implicit conversion to `{.a: i32}`.

`alias a = {.a: i32}.a` sounds like it creates a mutable reference to the struct member, but if the type has an implicit conversion to `{.a: i32}`, it would be a temporary copy for the call. As described, is it exposing the ability to mutate a field on the temporary generated by the cast? `alias` might be a problematic model for instance members on classes too; I think you might instead want a RefBind for `a = 0;`, but I'm not sure how `alias` distinguishes between a ValueBind versus RefBind use-case for an instance member.

Note, I'm not sure what the extended API for tuples would be: the API is very special. Should `extend api` on tuples be disallowed?

> * I don't think destructors have names, and so can't be aliased, and I don't see any situation in which some destructor could be called twice by the introduction of this feature. Perhaps you could clarify the concern with a concrete example where you think this would be a problem?

You don't comment on `extend base` being aliased, but I would note the risk there. It has a keyword name -- but maybe the base class shouldn't be aliased?

I think the destructor shouldn't be aliased, although https://github.com/carbon-language/carbon-lang/issues/2567 could give it a name.

Though thinking about this threw me into a tangent. If I `extend api` a type that has an `extend impl`, then it gets the interface functions but not the interface support, is that right?

Author: I have added text specifying this applies only to public members and does not allow access control modifiers. If we decide that `protected` is useful with `extend api`, we can add it in a separate proposal, but I currently am not aware of a use case. It would make more sense to me for `extend api` to treat protected members of the extended type like `extend base` does (still without specifying an access control keyword), but that is only useful if the extending type has access to the protected members of the extended type, which I don't see happening. For now, I prefer this simple rule until we have motivation to make it more complicated.

`alias a = {.a: i32}.a` by itself does not give a mutable reference, it just defines what the name `a` references. The result defines both value and reference binding operators, but the reference binding implementation requires a reference expression with type `{.a: i32}` that an implicit conversion to `{.a: i32}` would not provide.

SetField(&c, C.m, 42);
SetField(&c, C.n, 12);
Assert(GetField(c, C.m) == 42);
Assert(GetField(c, C.n) == 12);
```

### C++ pointer to member

In [the generic type of member section](#generic-type-of-a-class-member), the
names of members, such as `D.K`, `X.B`, `X.V`, and `C.n`, refer to zero-sized /
stateless objects where all the offset information is encoded in the type.
However, the definitions of `CallMethod`, `SetField`, and `GetField` do not
depend on that fact and will be usable with objects, such as C++
pointers-to-members, that include the offset information in the runtime object
state. So we can define binding implementations for them so that they may be
used with Carbon's `.(`**`)` and `->(`**`)` operators.

For example, this is how I would expect C++ code to call the above Carbon
functions:

```cpp
struct C {
  int F() const { return m + 1; }
  int m;
};

int main() {
  // pointer to data member `m` of class C
  int C::* p = &C::m;
  C c = {2};
  assert(c.*p == 2);
  assert(Carbon::GetField(c, p) == 2);
  Carbon::SetField(&c, p, 4);
  assert(c.m == 4);
  // pointer to method `F` of class C
  int (C::*q)() const = &C::F;
  assert(Carbon::CallMethod(&c, q) == 5);
}
```

> ### Member forwarding

Reviewer 3: It seems like `extend api` is primarily intended for a use-case that _also_ requires overloading `ValueBind` and `RefBind`. That sounds a bit like developers are provided `extend api` as part of a toolkit, but it doesn't stand on its own. Are there driving use-cases for `extend api` that wouldn't need additional wrappers? Maybe an alternative should be considered where `extend api` syntax implies callable bindings?

For example, a different option might be to say we support:

```
var ptr: T*;
extend api *ptr;
```

Regarding the note about implicit conversion below, perhaps slightly different syntax:

```
// Extend the pointer:
extend api *self.ptr;
// Extend the converted type:
extend api self as U;
```

Then, this could be equivalent to the full member forwarding syntax that is written out right now; the API of `typeof(*ptr)` is wrapped (not `alias`) onto the current type. I was considering expressing this as `let template Search:! auto = ptr->Search` but I think that's not quite correct due to the storage implication, which shouldn't exist.

Note for more custom behavior, `ValueBind` and `RefBind` could still be provided; I think the main difference is that there's a default behavior that results in `self: Self` working.

Author: `extend api T` works as-is with any type that can implicitly convert to `T`.

Reviewer 3: Maybe you can give some example use-cases where that's the desired use? I'd thought that would be more the edge case, as the implicit conversion would typically be a copy.

Reviewer 3: (from a perspective of syntax trade-offs, I would've thought that the implicit conversion case wouldn't be where `extend api` is needed)

Author: I've added an example using `extend api` with implicit conversion.

Reviewer: FWIW, I've added a comment below that may also be relevant -- I think it may make sense to reframe `extend api` as the basis functionality that other forms of `extend` build upon. That might make it a bit less surprising that it is rarely used on its own, and still useful to have as a primitive for complex cases.

    work by dereferencing the pointer `b.ptr`.

For the first ingredient, we need a way to customize
[simple member access](/docs/design/expressions/member_access.md) for the class
`Box(T)`. Normally simple member access on a value like `b`
[looks in the type of b](/docs/design/expressions/member_access.md#values).
However, types can designate other entities to also look up in using the
`extend` keyword (see [proposal #2760](/proposals/p2760.md#proposal)):

-   `extend impl as I` adds lookup into the implementation of an interface `I`.
-   `extend base: B` adds lookup into the base class `B`.
-   `extend adapt C` adds lookup into the adapted class `C`.
-   [mixins are also planning](/proposals/p2760.md#future-work-mixins) on using
    an `extend` syntax

The natural choice is to add another way of using `extend` with consistent
lookup rules as the other uses of `extend`, with conflicts handled as determined
by leads issues
[#2355](https://github.com/carbon-language/carbon-lang/issues/2355) and
[#2745](https://github.com/carbon-language/carbon-lang/issues/2745). I propose
`extend api T`:

```carbon
class Box(T:! type) {

>   var : T*;

Reviewer 3: ```suggestion
  var ptr: T*;
```

Is this missing the variable name? Looking at the examples below, it looks like it should be `ptr`

  extend api T;
}
```

This means that lookup into `Box(T)` also looks in the type `T` for members. In
this way, `b.Search` will find `b.(String.Search)`.

> **Note:** This is an alternative to defining an `alias` for each member of the
> extended type. This avoids having to repeat them, which is both lengthy and
> could get out of sync as the class evolves. The `extend api` approach works
> even if, as in this example, we don't know the names of the members of the

> > type being extended.

Reviewer 3: How is it intended to work with incomplete types? Consider the case:

```
class Box(T:! type, U:! type) {
  var t: T*;
  var u: U*;
  extend api T;
  extend api U;
}

class MyT;

class MyU {
  fn Foo();
}

// Should this be invalid since the API of Box cannot be established?
var b: Box(MyT, MyU);

fn Run1() {
  // Does this use MyU::Foo, or is it invalid because `T` is incomplete?
  b->Foo();
}

class MyT { fn Foo(); }

fn Run2() {
  // Does this use MyT::Foo because it now exists, or
  // MyU::Foo because that was the available API on instantiation?
  b->Foo();
}

```

Reviewer 2: I think at minimum it should be invalid to attempt a name lookup within a type that has not been defined. So:

```carbon
class U { fn G(); }
class Wrap(T:! type) {
  extend api T;
  extend api U;
  fn F[self: Self]();
}
class Undefined;
// Could be OK
fn F(w: Wrap(Undefined)) {
  // Could be OK
  w.F();
  // Should be error, because we perform lookup
  // for `G` in `T`, even though `U.G` exists.
  w.G();
}
```

Author: I've added a couple of sentences to clarify.

```
// Should this be invalid since the API of Box cannot be established?
var b: Box(MyT, MyU);
```
Valid since the API of Box is not being queried.

```
  // Does this use MyU::Foo, or is it invalid because `T` is incomplete?
  b->Foo();
```
Invalid because `T` is incomplete, and this triggers name lookup into it.

```
  // Does this use MyT::Foo because it now exists, or
  // MyU::Foo because that was the available API on instantiation?
  b->Foo();
```
I think this is a name conflict error, since it is found in both `MyT` and `MyU`. The API available on instantiation is not relevant.

Reviewer 3: I'm a little cautious of this, because it means we don't know the API of a type until the first call is done. Is there a use-case for allowing `extend api` on an incomplete type?

Reviewer 3: To try to make this a little more concrete, here's another example:

```
base class C {
  fn F();
}

class D {
  extend base: C;
  extend api B;
  fn G[self: Self]();
}

fn D.G[self: Self]() {
  // Does compilation fail here?
  self.F();
}
```

Specifically, although `F` has a definition on the base, it could also be defined on the api B. Since `extend api` is just aliasing things that don't have a conflict, and `B.F` wouldn't be a conflict (since `D` can shadow `base.F` right?), I think the name lookup here requires the incomplete type. But a compilation failure here might be a little less clear for developers.

Author: I've changed it to restrict to complete types.

Author: `extend api` is being moved to its own proposal.

-   Orthogonality: separating the binding process as a separate and independent
    step of using the members of a type.
-   Being consistent with our overall strategy for defining operators in terms
    of interface implementations.
-   Allows binding-related functionality to be define through
    [library APIs](/docs/project/principles/library_apis_only.md).
-   Increases uniformity by making member names into ordinary values with types.
-   Adds expressiveness, enabling member forwarding, passing a member as an
    argument, and other use cases.

These benefits advance Carbon's goals including:

-   [Language tools and ecosystem](/docs/project/goals.md#language-tools-and-ecosystem):
    by making it easier to reason about more Carbon entities within Carbon
    itself, and reducing the number of different concepts that have to be
    modeled.
-   [Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write):
    through increased consistency, uniformity, and expressiveness.
-   [Interoperability with and migration from existing C++ code](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code):

>     by adding supporti for pointer-to-member constructs.

Reviewer 3: ```suggestion
    by adding support for pointer-to-member constructs.
```

Notably, there is the process of
[instance binding](/docs/design/expressions/member_access.md#instance-binding),
that can convert a method into a bound method. This is described as an
uncustomizable process, with the members of classes being non-first-class names.

With
[proposal #3646: Tuples and tuple indexing](https://github.com/carbon-language/carbon-lang/pull/3646),
tuple indexing also uses the member-access syntax, except with numeric names for
the fields.

The currently accepted proposals for functions, most notably
[Proposal #2875: Functions, function types, and function calls](https://github.com/carbon-language/carbon-lang/pull/2875),
don't support all of the different function signatures for the `Call` interface.
For example, it does not support `addr self` or explicit compile-time
parameters. That is out of scope of this proposal, and will be addressed
separately, and means that `addr self` methods won't be considered here. The
difference between functions and methods, however, is in scope.

## Proposal

We propose that Carbon defines the compound member access operator, specifically

> `x.(y)`, to be defined as in terms of rewrites to invoking an interface method,

Reviewer 2: ```suggestion
`x.(y)`, to be defined in terms of rewrites to invoking an interface method,
```

like other operators. There are two different interfaces used, depending on
whether `x` is a value expression or a reference expression:

```carbon
// For value `x` with type `T` and `y` of type `U`,
// `x.(y)` is `y.((U as ValueBind(T)).Op)(x)`
interface ValueBind(T:! type) {
  let Result:! type;
  fn Op[self: Self](x: T) -> Result;
}

// For reference expression `var x: T` and `y` of type `U`,
// `x.(y)` is `*y.((U as RefBind(T)).Op)(&x)`
interface RefBind(T:! type) {
  let Result:! type;
  fn Op[self: Self](p: T*) -> Result*;
}
```

The other member access operators -- `x.y`, `x->y`, and `x->(y)` -- are defined
by how they rewrite into the `x.(y)` form using these two rules:

-   `x.y` is interpreted using the existing
    [member resolution rules](/docs/design/expressions/member_access.md#member-resolution).
    For example, `x.y` is treated as `x.(T.y)` for non-type values `x` with type
    `T`. These rules are expanded to support `extend api T` declarations
    described next.
-   `x->y` is interpreted as `(*x).y`.

In addition, we propose adding a declaration in a class body to find names

> defined in another class, `extend api T`:

Reviewer 2: This seems a bit narrower than necessary. Can we expand this beyond classes to also permit use in the definition of an interface or named constraint, and allow `T` to be any type? For example, given:

```carbon
interface A {
  fn F();
}
interface B {
  extend api A;
  fn G();
}
```

... `B.F` would be an alias for `A.F`, but like with plain `extend A;`, there would be no implied `require Self impls A`. So `extend I;` would then be equivalent to `require Self impls I; extend api I;`.

Similarly, for extension methods:

```carbon
interface A {
  fn F();
}
class C(T:! type) {
  extend api A;
}
impl C(i32) as A { fn F() { ... } }
```

Here, `C(i32).F()` would be valid, and `C(i64).F()` would fail because `C(i64)` doesn't implement `A`.

Finally:

```carbon
class Helpers {
  fn F[T:! type, self: T]() { DoStuffWith(self); }
  // ...
}
interface Iface {
  extend api Helpers;
  fn G();
  // ...
}
class C {
  extend impl as Iface;
}
fn Test(c: C) {
  // Calls Helpers.F from extended class Helpers
  c.F();
}
```

(This last example can be achieved with `final` functions in the interface instead, so it's probably not very motivating, but I think it demonstrates that there's a consistent rule we could use here.)

Author: Added, with a new section describing different cases.

```carbon
class Extended {
  fn F[self: Self]();
  fn G[self: Self]();
}

class Extending {
  fn G[self: Self]();
  extend api Extended;
  impl as ImplicitAs(Extended);
}

var e: Extending;
```

This means that lookup into `Extending` also looks in the type `Extended` for
members. In this way, `e.F` will find `e.(Extended.F)`. Note that does not make
`e.F()` actually work unless `Extended.F` was somehow legal to call on an value
of type `Extending`. In the example, this is accomplished by defining an
implicit conversion from `Extending` to `Extended`.

`extend api T` is equivalent to defining an `alias` for every member of `T` that

> doesn't have a conflicting name. This means `Extending` is equivalent to:

Reviewer 2: Given:

```carbon
class A {
  fn F();
}
class B {
  extend api A;
  // Hides A.F
  fn F();
}
class C {
  extend api A;
  extend api B;
}
```

... do we want `C.F` to be ambiguous, or to unambiguously name `B.F` (because it intends to hide `A.F`)? In C++, the corresponding rule for member name lookup in an inheritance hierarchy makes this ambiguous if `A` is a non-virtual base, but unambiguously picks `C.F` if `A` is a virtual base, and `extend api` feels a bit more like a virtual base situation to me, but I could go either way here.

Author: My opinion:
* For now we have a simple rule that says we look in each `extend`, and if a name is found multiple times it is an ambiguity error that needs qualification to disambiguate. I've added a sentence to say this.
* We could add a more complicated rule in the future, based on need, but the extra complexity doesn't seem justified given how rare I expect conflicting multiple extends to be. Would you like me to write an "alternative considered"?

For the specific example shown, the ambiguity is fixed by dropping `extend api A;` from `C`, but I assume that fix would not apply in other examples.

Reviewer 2: I think the simple rule is good for now. And it seems to leave space for us to reconsider in the future if we get evidence that a more complex rule is justified. I think this is worth making a brief "Alternative considered" note of.

Author: Will do in the follow-up `extend api` proposal

the type of binding that member with either a `C` value or variable.

```carbon
class __TypeOf_C_F {}
let __C_F:! __TypeOf_C_F = {};
class __ValueBind_C_F {
  adapt C;
}
class __RefBind_C_F {
  adapt C;
}

// and similarly for Static
```

These are the types that result from
[instance binding](/docs/design/expressions/member_access.md#instance-binding)
an instance of `C` with these member names. They define the bound method value
and bound method type of [proposal #2875](/proposals/p2875.md#bound-methods).
For example,

```carbon
let v: C = {.x = 3};
Assert(v.F() == 8);
Assert(v.Static() == 2);
var r: C = {.x = 4};
Assert(r.F() == 9);
Assert(r.Static() == 2);
```

is interpreted as:

```carbon
let v: C = {.x = 3};
Assert((v as __ValueBind_C_F).(Call(()).Op)() == 8);
Assert((v as __ValueBind_C_Static).(Call(()).Op)() == 2);
var r: C = {.x = 4};
Assert((r as __RefBind_C_F).(Call(()).Op)() == 9);
Assert((r as __RefBind_C_Static).(Call(()).Op)() == 2);
```

How does this arise?

1. First the simple member access is resolved using the type of the receiver: \
   `v.F` -> `v.(C.F)`, `v.Static` -> `v.(C.Static)`, `r.F` -> `r.(C.F)`, `r.Static`

>    -> `r.(C.Static)`. \

Reviewer 2: The current wording in `member_access.md` says that `v.(C.Static)` and `r.(C.Static)` are both invalid, because they don't perform member name lookup, instance binding, nor impl lookup -- the `v.` and `r.` portions are redundant. Is that rule removed by this proposal (meaning we don't enforce that constraint any more), or do we have some way to still provide that restriction?

Author: Added a note that this is now best effort.

Author: That rule is now removed by this proposal.

interface RefBind(T:! type) {
  let Result:! type;
  fn Op[self: Self](p: T*) -> Result*;
}
```

These binding operations are how we get from `__C_F` with type `__TypeOf_C_F` to
`v as __ValueBind_C_F` or `r as __RefBind_C_F`:

```carbon
impl __TypeOf_C_F as ValueBind(C)
    where .Result = __ValueBind_C_F {
  fn Op[self: Self](x: C) -> __ValueBind_C_F {
    return x as __ValueBind_C_F;
  }
}

impl __TypeOf_C_F as RefBind(C)
    where .Result = __RefBind_C_F {
  fn Op[self: Self](p: C*) -> __RefBind_C_F* {
    return p as __RefBind_C_F*;
  }
}
```

The last ingredient is the implementation of the call interfaces for these bound
types.

```carbon
// Since `C.F` takes `self: Self` it can be used with both
// value and reference expressions:
impl __ValueBind_C_F as Call(()) with .Result = i32;
impl __RefBind_C_F as Call(()) with .Result = i32;

// `C.Static` works the same as `C.F`, except it also
// implements the call interfaces on `__TypeOf_C_Static`.
// This allows `C.Static()` to work, in addition to
// `v.Static()` and `r.Static()`.
impl __ValueBind_C_Static as Call(()) with .Result = i32;
impl __RefBind_C_Static as Call(()) with .Result = i32;

> impl __TypeOf_C_Static as Call(()) where .Result = i32;

Reviewer 2: Interesting. I'd been expecting for `__TypeOf_C_Static as ValueBind(C)` to have a `.Result` of `__TypeOf_C_Static`. But I suppose that doesn't work for `RefBind`, because we don't have a `__TypeOf_C_Static*` to return.

This approach also means that `v.(v.(C.Static))` is rejected, which seems like a good property.

Author: Should I add an alternative considered?

Author: Added alternative in https://github.com/carbon-language/carbon-lang/pull/3720/commits/fd5815d52b1199072e98fc4b54e5f6f28a0e76d4

```

**Note**: We could reduce the number of implementations required by making
`__RefBind_*` an
[extending adapter](/docs/design/generics/details.md#extending-adapter) of the
corresponding `__ValueBind_*` class.

### Inheritance and other implicit conversions

Now consider methods of a base class:

```carbon
base class B {
  fn F[self: Self]();
  virtual fn V[self: Self]();
}

class D {
  extend base: B;
  impl fn V[self: Self]();
}

var d: D = {}
d.(B.F)();
d.(B.V)();
```

To allow this to work, we need the implementation of the bind interfaces to
allow implicit conversions:

```carbon
impl [T:! ImplicitAs(B)] __TypeOf_B_F as ValueBind(T)
    where .Result = __ValueBind_B_F {
  fn Op[self: Self](x: T) -> __ValueBind_B_F {
    return (x as B) as __ValueBind_B_F;
  }
}

impl [T:! type where .Self* impls ImplicitAs(B*)]
    __TypeOf_B_F as RefBind(T)
    where .Result = __RefBind_B_F {
  fn Op[self: Self](p: T*) -> __RefBind_B_F* {
    return (p as B*) as __RefBind_B_F*;
  }
}
```

> This matches the expected semantics method calls, even for methods of final

Reviewer 2: ```suggestion
This matches the expected semantics of method calls, even for methods of final
```

```carbon
// `C.G` will only bind to values that can implicitly convert
// to type `Different`.
impl [T:! ImplicitAs(Different)] __TypeOf_C_G as ValueBind(T)
    where .Result = __ValueBind_C_G;
```

### Data fields

The same `ValueBind` and `RefBind` operations allow us to define access to the
data fields in an object, without any additional changes.

For example, given a class with a data member `m` with type `i32`:

```carbon
class C {
  var m: i32;
}
```

we want the usual operations to work, with `x.m` equivalent to `x.(C.m)`:

```carbon
let v: C = {.m = 4};
var x: C = {.m = 3};
x.m += 5;
Assert(x.(C.m) == v.m + v.(C.m));
```

To accomplish this we will, as before, associate an empty (stateless or
zero-sized) type with the `m` member of `C`, that just exists to support the
binding operation. However, this time the result type of binding is simply
`i32`, the type of the variable, instead of a new, dedicated type.

```carbon
class __TypeOf_C_m {}
let __C_m:! __TypeOf_C_m = {};

impl __TypeOf_C_m as ValueBind(C) where .Result = i32 {
  fn Op[self: Self](x: C) -> i32 {
    // Effectively performs `x.m`, but without triggering binding again.

>     return value_compiler_intrinsic(v, __OffsetOf_C_m, i32)

Reviewer 2: ```suggestion
    return value_compiler_intrinsic(x, __OffsetOf_C_m, i32)
```

  }
}

impl __TypeOf_C_m as RefBind(C) where .Result = i32 {
  fn Op[self: Self](p: C*) -> i32* {
    // Effectively performs `&p->m`, but without triggering binding again,
    // by doing something like `((p as byte*) + __OffsetOf_C_m) as i32*`
    return offset_compiler_intrinsic(p, __OffsetOf_C_m, i32);
  }
}
```

These definitions give us the desired semantics:

```carbon
// For value `v` with type `T` and `y` of type `U`,

> // `v.(y)` is `y.(U as ValueBind(T)).Op(v)`

Reviewer 2: ```suggestion
// `v.(y)` is `y.((U as ValueBind(T)).Op)(v)`
```

v.m == v.(C.m)
    == v.(__C_m)
    == v.(__C_m as (__TypeOf_C_m as ValueBind(C)))

>     == __C_m.(__TypeOf_C_m as ValueBind(C)).Op(v)

Reviewer 2: ```suggestion
    == __C_m.((__TypeOf_C_m as ValueBind(C)).Op)(v)
```

>     == value_compiler_intrinsic(v, __OffsetOf_C_m, i32)

Reviewer 2: We can go deeper:

```
    == (__TypeOf_C_m as ValueBind(C)).Op.((__TypeOf_ValueBindFacet_Op as ValueBind(__TypeOf_C_m)).Op))(__C_m)(v)
```

That is:

```carbon
let bind1: auto = (typeof(C.m) as ValueBind(C)).Op;
let bind2: auto = (typeof(bind1) as ValueBind(typeof(C.m)).Op;
v.m == bind1.(bind2)(C.m)(v)
```

... so some kind of currying of the `self` parameter sort of falls out. (We can iterate the transform more times, but it stops being interesting -- you get `bind1.(bind2)(bind3)(...)(bindN)(C.m)(v)`.)

I guess the main takeaway here is that at some point we need to recognize that we have a "real" method on the right-hand side of a `.(...)` and a compatible `self` on the left-hand side, and perform the transform to `value_compiler_intrinsic` rather than iteratively rewriting forever.

Should there be some limit on the number of times we can rewrite? Maybe the limit should be 1, so that we don't actually rewrite `x.(y)` as `y.(bind_op)(x)`, but instead rewrite it as `bind_compiler_intrinsic(y, bind_op)(x)`? `bind_op` should always be an actual function that we just plucked out of an `impl`.

Author: I wrote a short note choosing the limit of 1 option in the first "Details" section, after some reorganization and exposition.

  int C::* p = &C::m;
  C c = {2};
  assert(c.*p == 2);
  assert(Carbon::GetField(c, p) == 2);
  Carbon::SetField(&c, p, 4);
  assert(c.m == 4);
  // pointer to method `F` of class C
  int (C::*q)() const = &C::F;
  assert(Carbon::CallMethod(&c, q) == 5);
}
```

### Member forwarding

Consider a class that we want to act like it has the members of another type.
For example, a type `Box(T)` that has a pointer to a `T` object allocated on the
heap:

```carbon
class Box(T:! type) {
  var ptr: T*;
  // ???
}
```

`Box(T)` should act like it has all the members of `T`:

```carbon
class String {
  fn Search[self: Self](c: u8) -> i32;
}
var b: Box(String) = ...;
var position: i32 = b.Search(32);
```

There are two ingredients to make this work:

-   We need some way to make `b.Search` be equivalent to `b.(String.Search)` not
    `b.(Box(String).Search)`.
-   We need the act of binding a method of `String` to a `Box(String)` value
    work by dereferencing the pointer `b.ptr`.

For the first ingredient, we need a way to customize
[simple member access](/docs/design/expressions/member_access.md) for the class
`Box(T)`. Normally simple member access on a value like `b`

> [looks in the type of b](/docs/design/expressions/member_access.md#values).

Reviewer 2: ```suggestion
[looks in the type of `b`](/docs/design/expressions/member_access.md#values).
```

> example, a class could extend the API of a class it implicitly converts to,
> see
> [inheritance and other implicit conversions](#inheritance-and-other-implicit-conversions).
> Or a class could have members specifically intended for use by another class,
> as
> [extension methods](https://github.com/carbon-language/carbon-lang/issues/1345)
> -- effectively acting as mixin except it can't add member variables.

For the second ingredient, the binding interfaces already provide everything
needed, we just need to make a parameterized implementation of them:

```carbon
impl forall [T:! type, U:! ValueBind(T)]
    U as ValueBind(Box(T)) where .Result = U.Result {
  fn Op[self: Self](x: Box(T)) -> Result {
    // Uses `ValueBind(T).Op` based on type of `U`.
    return self.Op(*x.ptr);

    // NOT `return x.ptr->(self)` since that attempts
    // to do reference binding not value binding.
  }
}

impl forall [T:! type, U:! RefBind(T)]
    U as RefBind(Box(T)) where .Result = U.Result {
  fn Op[self: Self](p: Box(T)*) -> Result* {
    return self.Op(p->ptr);
    // Or equivalently:
    // return p->ptr->(self);
  }
}

> ```

Reviewer 2: As an aside:

This `Box` is interesting, in that it preserves the expression category -- you can only call mutating operations on a `Box` reference expression, not on a `Box` value expression. It strikes me that we can use this to build something even more like a C++ reference; the trick is to permit value binding for anything where `T` permits *reference* binding:

```
class Ref(T:! type) {
  var ptr: T*;
  extend api T;
}
impl forall [T:! type, U:! RefBind(T)]
    U as ValueBind(Ref(T)) where .Result = Ref(U.Result) {
  fn Op[self: Self](p: Ref(T)) -> Ref(Result) {
    return {.ptr = self.Op(p->ptr)};
  }
}
// (Plus impls for the other three combinations of ValueBind and RefBind.)
impl forall [S:! type, T:! Call(S)] Ref(T) as Call(S) where .Result = T.Result {
  // ...
}

class C {
  fn F[addr self: Self*]();
  var n: i32;
}
fn F(r: Ref(C)) {
  // OK, even though `r` is a value expression.
  // `r.F` is a `Ref(<bound member function>)`,
  // which is callable.
  r.F();

  // OK, forms a reference to the `n` member.
  let n_ref: Ref(i32) = r.n;
}
```

(But of course, `Ref(T)` doesn't preserve the interface implementations of `T`.)

It's both comforting and a little unnerving that that seems to work very well.

Author: Should I include any text about this in the proposal?

Reviewer: If factoring out the `extend api`, it seems worth doing a little bit of exploration of what a library-reference-type would look like. It particularly seems interesting to compare and contrast `Box(T)` and `Ref(T)`.

Author: Added a note to look at this in the successor proposal.

A few observations:

-   These declarations use `Box` to satisfy the orphan rule, and so this
    approach only works with `Box(T)` values, not types that can implicitly
    convert to `Box(T)`.
-   The implementation of the `Op` method is where we follow the pointer member
    `ptr` of `Box(T)` to get a `T` value that is compatible with the member
    being bound. This resolves the type mismatch that is introduced by allowing
    name resolution to find another type's members.
-   We have to be a little careful in the implementation of `ValueBind(Box(T))`
    to still use value binding even when we get a reference expression from
    dereferencing the pointer `ptr`.

With these two ingredients, `b.Search(32)` is equivalent to
`b.(String.Search)(32)`, which is then equivalent to
`b.ptr->(String.Search)(32)` or `b.ptr->Search(32)` (since `b.ptr` has type
`String*`).

> **Future:** We might want `extend api` to also get (some) interface
> implementations, like `extend base` does.

> ### Future: tuple indexing

Reviewer 2: On the subject of future things: how do interface members fit into this picture? I think the name of an interface member probably needs a fairly interesting relationship with `ValueBind` / `RefBind` that performs impl lookup and potentially also instance binding.

Author: Added a section. I had to add `TypeBind` to get this to work with non-instance members.

is a value expression, a reference expression, or a facet:

```carbon
// This determines the type of the result of binding. It is a
// separate interface shared by `ValueBind` and `RefBind` to
// ensure they produce the same result type. We don't want the
// type of an expression to depend on the expression category
// of the arguments.
interface Bind(T:! type) {
  let Result:! type;
}

// For a value expression `x` with type `T` and an expression
// `y` of type `U`, `x.(y)` is `y.((U as ValueBind(T)).Op)(x)`
interface ValueBind(T:! type) {
  extend Bind(T);
  fn Op[self: Self](x: T) -> Result;
}

// For a reference expression `x` using a binding `var x: T`
// and an expression `y` of type `U`, `x.(y)` is
// `*y.((U as RefBind(T)).Op)(&x)`
interface RefBind(T:! type) {
  extend Bind(T);
  fn Op[self: Self](p: T*) -> Result*;
}

// For a facet value, which includes all type values, `T` and
// an expression `y` of type `U`, `T.(y)` is
// `y.((U as TypeBind(T)).Op)()`.
interface TypeBind(T:! type) {

>   let Result:! type;

Reviewer 2: Are there coherence concerns here? For a given `x.(y)`, we might not know whether `x` is a facet value.

I don't think we can use `Bind` here, because we'd want to require that if `M impls TypeBind(T) with .Result = R` then `M impls Bind(type) with .Result = R` (and in general, `forall [F is a facet type] M impls Bind(F) with .Result = R`). But `R` needs to depend on `T`, and `T` isn't in scope in that impl of `Bind`.

I suppose the upshot of this is that the "facet on the left of `.`" check must be done in the definition of the generic, and if it turns out that the left-hand expression is a facet once we know its type, that's just too late, and you get the `ValueBind` / `RefBind` behavior instead. Something similar comes up elsewhere, where we say a direct member access into a facet doesn't use an overloadable binding operator.

How much should we be worried about this?

Author: My lack of worry mostly comes from not having an alternative to compare it to.

Author: Added a sentence to the note about `BindToType`.

> [non-instance interface members](#non-instance-interface-members).

Those implementations are how we get from `__C_F` with type `__TypeOf_C_F` to
`v as __Bind_C_F` or `&r as __Bind_C_F*`, conceptually following these steps:

```carbon
// `v` is a value and so uses `ValueBind`
v.F() == v.(C.F)()
      == v.(__C_F)()
      == __C_F.((__TypeOf_C_F as ValueBind(C)).Op)(v)()
      == (v as __Bind_C_F)()

// `r` is a reference expression and so uses `RefBind`
r.F() == r.(C.F)()
      == r.(__C_F)()
      == (*__C_F.((__TypeOf_C_F as RefBind(C)).Op)(&r))()
      == (*(&r as __Bind_C_F*))()
```

However, to avoid recursive application of these same rules, we need to avoid
expressing this in terms of evaluating `__C_F.(`...`)`. Instead the third step
shall instead use an intrinsic compiler primitive, as in:

```carbon
// `v` is a value and so uses `ValueBind`
v.F() == v.(C.F)()
      == v.(__C_F)()
      == inlined_method_call_compiler_intrinsic(
              <function body (__TypeOf_C_F as ValueBind(C)).Op overload 0>,
              __C_F, (v))()
      == (v as __Bind_C_F)()

// `r` is a reference expression and so uses `RefBind`
r.F() == r.(C.F)()
      == r.(__C_F)()

>       == (*inlined_method_call_compiler_intrinsic(

Reviewer 2: This makes me wonder: does this need to be a primitive dereference, rather than this desugaring to [a call to `Pointer.Dereference`](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/values.md#dereferencing-customization)? I think not, because we'd end up in `ValueBind` rather than `RefBind`; do you agree?

I think we shouldn't use a primitive dereference here unless we really need to: while there might still be an argument for avoiding the extra work of an impl lookup for the `*` here, if we want to make `*` bypass implementation lookup for pointers, we may want to do that centrally in the rewrite for `*` rather than only in this one case. Also, as a future extension, we might want to allow `RefBind.Op` to return a value that implements `Pointer where .ValueT == Result` rather than requiring it to return exactly `Result*`, at which point we'd need the full dereference machinery here.
